moonraker/components/authorization.py

```diff
@@ -356,16 +388,71 @@ async def _handle_password_reset(self,
             'action': "user_password_reset"
         }
 
+    def _login_ldap_user(self, username, password) -> bool:
+        base_dn = str(self.ldap_base_dn)
+        client = LDAPClient(self._generate_ldap_url_(str(self.ldap_server)))
+        client.set_credentials("SIMPLE", self.ldap_bind_dn,
+                               self.ldap_bind_password)
+        client.set_cert_policy("allow")
+        conn = LDAPConnection
+        try:
+            conn = client.connect(is_async=False, timeout=10)
+            ldap_filter = ("(&(objectClass=Person)(%s=" + '%s))') % \
+                          ("sAMAccountName"
+                           if self.ldap_type_ad
+                           else "uid",
+                           username)
+            user = conn.search(
+                base_dn,
+                LDAPSearchScope.SUBTREE,
+                ldap_filter,
+                ['memberOf']
+            )
+            auth_username = str(user[0]['DN'])
+            client.set_credentials("SIMPLE", auth_username, password)
+            conn = client.connect(is_async=False, timeout=10)
+            conn.close()
+            if self.ldap_group_dn is None:
+                return True
+            if len(user[0]['memberOf']) > 0:
+                for group in user[0]['memberOf']:
+                    if str(group) == str(self.ldap_group_dn):
+                        return True
+        except ConnectionError:
+            return False
+        except AuthenticationError:
+            conn.close()
+            return False
+        except TimeoutError:
+            return False
+        conn.close()
+        return False
+
     def _login_jwt_user(self,
                         web_request: WebRequest,
                         create: bool = False
                         ) -> Dict[str, Any]:
         username: str = web_request.get_str('username')
         password: str = web_request.get_str('password')
+        source: str = web_request.get_str('source', self.default_source).lower()
         user_info: Dict[str, Any]
         if username in RESERVED_USERS:
             raise self.server.error(
                 f"Invalid Request for user {username}")
+        if source == "ldap":
+            if self.ldap_server is None or self.ldap_base_dn is None \
+                    or self.ldap_group_dn is None \
+                    or self.ldap_bind_password is None \
+                    or self.ldap_bind_dn is None:
+                raise self.server.error(
+                    "ldap: Configuration is not given", 401
+                )
```

Comment:
It think it would  be better to move this code inside `login_ldap_user` and raise all authentication errors there.   This allows us to simply this method and simply call:
```python
if source == "ldap":
    await self._login_ldap_user(username, password)
    if username not in self.users:
        create = True
```

One note:  We will need to make `_login_jwt_user()` an async method and await its result in the two places its called (those handlers are already async methods).  `_login_ldap_user()` also needs to be an async method so we can call use the async variant of bonsai's ldap client.